File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/main/java/com/ntnu/gidd/security/filter/JWTUsernamePasswordAuthenticationFilter.java
package com.ntnu.gidd.security.filter;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.ntnu.gidd.controller.request.LoginRequest;
import com.ntnu.gidd.security.UserDetailsImpl;
import com.ntnu.gidd.security.config.JWTConfig;
import com.ntnu.gidd.security.token.JwtToken;
import com.ntnu.gidd.security.token.JwtTokenFactory;
import com.ntnu.gidd.security.token.TokenFactory;
import com.ntnu.gidd.service.token.RefreshTokenService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;

import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;


/**
 * Filter to check the existence and validity of the access token on the Authorization header.
 */
@Slf4j
public class JWTUsernamePasswordAuthenticationFilter extends UsernamePasswordAuthenticationFilter {

    private RefreshTokenService refreshTokenService;
    private AuthenticationManager authenticationManager;
    private TokenFactory tokenFactory;
    private ObjectMapper objectMapper;
    private JWTConfig jwtConfig;

    public JWTUsernamePasswordAuthenticationFilter(RefreshTokenService refreshTokenService, AuthenticationManager authenticationManager, JWTConfig jwtConfig) {
        this.refreshTokenService = refreshTokenService;
        this.authenticationManager = authenticationManager;
        this.jwtConfig = jwtConfig;
        this.tokenFactory = new JwtTokenFactory(jwtConfig);
        this.objectMapper = new ObjectMapper();

        this.setRequiresAuthenticationRequestMatcher(
                new AntPathRequestMatcher(jwtConfig.getUri() + "/login", "POST"));
    }

    /**
     * Attempt to Authenticate the request.
     */
    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException {

        try {
            LoginRequest credentials = new ObjectMapper().
                    readValue(request.getInputStream(), LoginRequest.class);

            log.debug("Found credentials, authenticating user: {}", credentials);

            UsernamePasswordAuthenticationToken authentication = new UsernamePasswordAuthenticationToken(
                    credentials.getEmail(),
                    credentials.getPassword(),
                    new ArrayList<>());

            return authenticationManager.authenticate(authentication);
        } catch (IOException exception) {
            log.error("Exception occurred while authenticating user", exception);
            throw new RuntimeException(exception);
        }
    }

    /**
     * Create JWT access and refresh token  and write it to response header upon successful authentication.
     */
    @Override
    protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication auth) throws IOException {
        UserDetailsImpl userDetails = ((UserDetailsImpl) auth.getPrincipal());

        JwtToken accessToken = tokenFactory.createAccessToken(userDetails);
        JwtToken refreshToken = tokenFactory.createRefreshToken(userDetails);
        refreshTokenService.saveRefreshToken(refreshToken);
        
        log.debug("Successfully created access and refresh token for user (email:{})", userDetails.getUsername());

        Map<String, String> tokens = new HashMap<>();
        tokens.put("token", accessToken.getToken());
        tokens.put("refreshToken", refreshToken.getToken());

        response.addHeader(jwtConfig.getHeader(), jwtConfig.getPrefix() + accessToken);
        response.setContentType("application/json");
        response.setCharacterEncoding("UTF-8");
        objectMapper.writeValue(response.getWriter(), tokens);
        
        log.debug("Successfully authenticated user. (email:{})", userDetails.getUsername());
    }

}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/main/kotlin/ntnu/idatt2105/security/filter/JWTUsernamePasswordAuthenticationFilter.kt
package ntnu.idatt2105.security.filter

import com.fasterxml.jackson.databind.ObjectMapper
import ntnu.idatt2105.security.config.JWTConfig
import ntnu.idatt2105.security.dto.LoginRequest
import ntnu.idatt2105.security.service.RefreshTokenService
import ntnu.idatt2105.security.token.JwtAccessToken
import ntnu.idatt2105.security.token.JwtToken
import ntnu.idatt2105.security.token.JwtTokenFactory
import ntnu.idatt2105.security.token.TokenFactory
import ntnu.idatt2105.user.service.UserDetailsImpl
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import java.io.IOException
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class JWTUsernamePasswordAuthenticationFilter : UsernamePasswordAuthenticationFilter {

    private var tokenFactory: TokenFactory
    private var objectMapper: ObjectMapper
    private var refreshTokenService: RefreshTokenService?
    private var jwtConfig: JWTConfig

    constructor(refreshTokenService: RefreshTokenService, authenticationManager: AuthenticationManager, jwtConfig: JWTConfig) {
        this.refreshTokenService = refreshTokenService
        this.jwtConfig = jwtConfig
        this.authenticationManager = authenticationManager
        tokenFactory = JwtTokenFactory(jwtConfig)
        objectMapper = ObjectMapper()
        setRequiresAuthenticationRequestMatcher(
                AntPathRequestMatcher(jwtConfig.uri + "/login", "POST"))
    }

    @Throws(AuthenticationException::class)
    override fun attemptAuthentication(request: HttpServletRequest, response: HttpServletResponse?): Authentication? {
        return try {
            val credentials: LoginRequest = ObjectMapper().readValue(request.inputStream, LoginRequest::class.java)
            val authentication = UsernamePasswordAuthenticationToken(
                    credentials.email,
                    credentials.password,
                    ArrayList())
            authenticationManager.authenticate(authentication)
        } catch (exception: IOException) {
            throw RuntimeException(exception)
        }
    }

    @Throws(IOException::class)
    override fun successfulAuthentication(request: HttpServletRequest?, response: HttpServletResponse, chain: FilterChain?, auth: Authentication) {
        val userDetails: UserDetailsImpl = auth.getPrincipal() as UserDetailsImpl
        val accessToken: JwtAccessToken = tokenFactory.createAccessToken(userDetails)
        val refreshToken: JwtToken = tokenFactory.createRefreshToken(userDetails)
        refreshTokenService?.saveRefreshToken(refreshToken)
        val tokens: MutableMap<String, String> = HashMap()
        tokens.putIfAbsent("token", accessToken.getToken())
        tokens.putIfAbsent("refreshToken", refreshToken.getToken())

        response.addHeader(jwtConfig.header, "${jwtConfig.prefix} ${accessToken.getToken()}")
        response.contentType = "application/json"
        response.characterEncoding = "UTF-8"
        objectMapper.writeValue(response.writer, tokens)
    }
}
